[Feature][Multimodal] Implement LRU cache for multimodal embeddings

ZhengWG opened the pull request:
## Motivation

Current implementation for embedding cache only caches for one request. 

However, in some cases, same requests may have different context data but the same multimodal data.

 Keeping previous embedding cache can significantly speed up the embedding forward process by avoiding redundant computation of identical multimodal inputs.

## Modifications

Implement LRU-based mm_cache: only free the last recently used embedding cache when cache size is not enough.

Test the evict func:

```bash
python test/srt/models/test_vlm_models.py
```

Test result:

<img width="2362" height="1150" alt="image" src="https://github.com/user-attachments/assets/1aeedd4c-07d9-440a-a860-6978a2dca26b" />


## Checklist

- [ ] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

ZhengWG committed:
feat: implement LRU cache for multimodal embeddings and optimize memory management

JustinTong0323 committed:
Merge branch 'main' into optimize-embedding-cache

JustinTong0323 committed:
Merge branch 'main' into optimize-embedding-cache

JustinTong0323 committed:
fix lint

Signed-off-by: Xinyuan Tong <xinyuantong.cs@gmail.com>

JustinTong0323 commented on python/sglang/srt/managers/mm_utils.py:
```diff
@@ -269,20 +269,12 @@ def _get_chunked_prefill_embedding(
                     "`SGLANG_VLM_CACHE_SIZE_MB` environment variable."
                 )
 
```
```
if not embedding_cache.put(embedding_items_hash, embedding_per_req):
    print_warning_once(
        "Multimodal embedding cache is full. Consider increasing the "
        "`SGLANG_VLM_CACHE_SIZE_MB` environment variable."
    )
```
This scenario only occurs when a single embedding exceeds the cache size, and perhaps this detail can be clarified here.

ZhengWG replied:
Thank you for the view! Please help check the updated warning message.

JustinTong0323 commented on python/sglang/srt/mem_cache/multimodal_cache.py:
```diff
@@ -28,7 +43,12 @@ def has(self, mm_hash: int) -> bool:
+            return self.mm_cache[mm_hash]
+        return None
 
     def free(self, mm_hash: int) -> bool:
```
This method is not used and can likely be deleted.

JustinTong0323 commented on python/sglang/srt/mem_cache/multimodal_cache.py:
```diff
@@ -1,24 +1,39 @@
+
     def put(self, mm_hash: int, embedding: torch.Tensor) -> bool:
         if mm_hash in self.mm_cache:
+            # Move to end (most recently used) and update
```
I'm curious about the logic here: it seems not possible to cache an embedding with the same hash, given the logic in mm_utils:
```
        embedding_items_hash = get_embedding_hash(embedding_items_per_req)
        # if all items has been prefixed, we do not need to calculate embedding
        if all([offset_end < prefix_length[i] for _, offset_end in items_offset]):
            continue
        embedding_per_req = embedding_cache.get(embedding_items_hash)
        if embedding_per_req is None:
            embedding_per_req = data_embedding_func(embedding_items_per_req)
            if not embedding_cache.put(embedding_items_hash, embedding_per_req):
                print_warning_once(
                    "Multimodal embedding cache is full. Consider increasing the "
                    "`SGLANG_VLM_CACHE_SIZE_MB` environment variable."
                )
```

ZhengWG replied:
You're right. It's not necessary to verify here."

JustinTong0323 commented:
Thank you for your contribution! This is a really great feature.

ZhengWG committed:
fix: clean redundant logic && enhance warning message

JustinTong0323 committed:
Merge branch 'main' into optimize-embedding-cache

JustinTong0323 commented:
@ZhengWG  Workload of current CI is not able to fill all the mm cache and test the logic of evict, may you help add one test? Thanks!
（Maybe set the cache size to a small one, like 5M）

ZhengWG committed:
CI: add test for embedding evict logic

ZhengWG commented:
> @ZhengWG Workload of current CI is not able to fill all the mm cache and test the logic of evict, may you help add one test? Thanks! （Maybe set the cache size to a small one, like 5M）

Sorry for the late reply. I’ve added cache evict test—could you please check if the changes meet your expectations?

JustinTong0323 commented:
> > @ZhengWG Workload of current CI is not able to fill all the mm cache and test the logic of evict, may you help add one test? Thanks! （Maybe set the cache size to a small one, like 5M）
> 
> Sorry for the late reply. I’ve added cache evict test—could you please check if the changes meet your expectations?

Thank you so much! It would be wonderful if you could carry out some additional offline tests and paste the results under this PR to demonstrate the behavior of the LRU Cache. For instance, printing a log whenever an eviction occurs would be cool!

JustinTong0323 commented:
Also please format the codes with pre-commit

ZhengWG committed:
Apply pre-commit formatting to test_vlm_models.py

ZhengWG committed:
Add eviction detection logic to VLM test with log capture and analysis

ZhengWG commented:
> > > @ZhengWG Workload of current CI is not able to fill all the mm cache and test the logic of evict, may you help add one test? Thanks! （Maybe set the cache size to a small one, like 5M）
> > 
> > 
> > Sorry for the late reply. I’ve added cache evict test—could you please check if the changes meet your expectations?
> 
> Thank you so much! And it would be wonderful if you could carry out some additional offline tests and paste the results under this PR to demonstrate the behavior of the LRU Cache. For instance, printing a log whenever an eviction occurs would be cool!

I’ve added debug-level logs for eviction events (to avoid performance impact) and CI-test logic to detect them. Test results are included in the PR description. 
Could you help double-check the changes? Thanks!

ZhengWG committed:
fix: remove redundant logic

JustinTong0323 committed:
Merge branch 'main' into optimize-embedding-cache

JustinTong0323 committed:
Merge branch 'main' into optimize-embedding-cache

ZhengWG committed:
fix: redirect logs to avoid hang in CI

ZhengWG committed:
Merge branch 'main' into optimize-embedding-cache

ZhengWG committed:
fix: remove redundant code in CI

ZhengWG committed:
Merge branch 'main' into optimize-embedding-cache

JustinTong0323 committed:
Merge branch 'main' into optimize-embedding-cache

JustinTong0323 committed:
Merge branch 'main' into optimize-embedding-cache

JustinTong0323 committed:
Merge branch 'main' into optimize-embedding-cache
